### PR TITLE
Mitigate BREACH attack

### LINKED ISSUE #64: Mask CSRF tokens to mitigate BREACH attack
See the following pull request: https://github.com/rails/rails/pull/11729. Is there anybody interested in giving a shot at it? I can try, but I'm probably not going to get to it unless next week.

### PR DESCRIPTION
**Pull request by jkowens**
Implement the Rails solution to mitigate BREACH attack. See rails/rails/pull/11729 and rails/rails/pull/16570. Resolves #64.

To actually mitigate the attack forms and meta tags will need to use a masked authenticity token. This can be retrieved by using the following:

```
Rack::Protection::AuthenticityToken.token(session)
```

In a hidden input:

```
<input type="hidden" name="authenticity_token" value="#{Rack::Protection::AuthenticityToken.token(session)}" />
```

Or in a meta tag:

```
<meta name="csrf-token" content="#{Rack::Protection::AuthenticityToken.token(session)" />
```

The `#token` method can also be called on any of the classes that extend `AuthenticityToken` such as `FormToken`, and `RemoteToken`.

Since this uses the same algorithm to mask tokens as Rails, this update will fix compatibility with with Rails >= 4.2.0.

### CODE REVIEW

**Comment by zzak** [comment]
Haha nice! I was actually hacking together a sloppy version of this myself, glad you got to it first :)

**Comment by jkowens** [comment]
Haha sorry! Feel free to go with what you've got started instead.

**Comment by zzak** [comment]
@jkowens Well I'd like to merge sinatra/sinatra#1167 first, could you submit your patch upstream after that happens?

My patch isn't ready, and if this works, I'd rather use that then spend more time on it :)

File: lib/rack/protection/authenticity_token.rb [hunk 1]
```diff
@@ -20,15 +22,91 @@ class AuthenticityToken < Base
       default_options :authenticity_param => 'authenticity_token',
                       :allow_if => nil
 
+      TOKEN_LENGTH = 32
```
Comment (zzak) [inline lib/rack/protection/authenticity_token.rb]: This should be default option, which can be configurable, call it `:authenticity_token_length`

**Comment by jkowens** [comment]
@zzak Sure thing, I'll get this submitted upstream. I've refactored the code a bit and made the authenticity token length an option. I added a `Rack::Protection::Utils` module to help accomplish that so hopefully that's ok.

**Comment by zzak** [comment]
A new module is ok, but I think `Base` is also fine! There's already a `random_string` method which was previously used for generating tokens.

**Comment by zzak** [comment]
@jkowens Looking forward to your patch :)

**Pull request closed without merging**